File: content-solutions/heap/course-plan.java
import java.util.*;

class Solution {
    public int solution(int[][] courses) {
        int[][] sorted = courses.clone();
        Arrays.sort(sorted, (a, b) -> Integer.compare(a[1], b[1]));
        PriorityQueue<Integer> h = new PriorityQueue<>(Collections.reverseOrder());
        long time = 0;
        for (int[] c : sorted) {
            h.offer(c[0]);
            time += c[0];
            if (time > c[1]) time -= h.poll();
        }
        return h.size();
    }
}

File: content-solutions/heap/course-plan.py
import heapq


def solution(courses):
    h = []
    time = 0
    for d, last in sorted(courses, key=lambda c: c[1]):
        heapq.heappush(h, -d)
        time += d
        if time > last:
            time -= -heapq.heappop(h)
    return len(h)
